ibis/expr/signature.py

```diff
@@ -165,48 +166,63 @@ def __new__(metacls, clsname, bases, dct):
 
         return super().__new__(metacls, clsname, bases, attribs)
 
+    def __call__(cls, *args, **kwargs):
+        bound = cls.__signature__.bind(*args, **kwargs)
+        bound.apply_defaults()
+
+        kwargs = {}
+        for name, value in bound.arguments.items():
+            param = cls.__signature__.parameters[name]
+            kwargs[name] = param.validate(kwargs, value)
 
-class Annotable(metaclass=AnnotableMeta):
+        instance = super().__call__(**kwargs)
+        instance.__post_init__()
+        return instance
+
+
+class Annotable(Hashable, metaclass=AnnotableMeta):
     """Base class for objects with custom validation rules."""
 
-    __slots__ = ("_args",)
+    __slots__ = ("args", "_hash")
 
-    def __init__(self, *args, **kwargs):
-        bound = self.__signature__.bind(*args, **kwargs)
-        bound.apply_defaults()
-        for name, value in bound.arguments.items():
-            param = self.__signature__.parameters[name]
-            setattr(self, name, param.validate(self, value))
-        self._validate()
-        self._args = None  # materialize _args lazily
+    def __init__(self, **kwargs):
+        for name, value in kwargs.items():
+            object.__setattr__(self, name, value)
 
-    def _validate(self):
-        pass
+    # TODO(kszucs): split for better __init__ composability but we can
+    # directly call it from __init__ too
+    def __post_init__(self):
+        args = tuple(getattr(self, name) for name in self.argnames)
+        object.__setattr__(self, "args", args)
+        object.__setattr__(self, "_hash", hash(args))
 
     def __eq__(self, other) -> bool:
         if self is other:
             return True
+        # TODO(kszucs): measure using the precomputed hash
+        # (
+        #     type(self) == type(other) and
+        #     self._hash == other._hash and
+        #     self.args == other.args
+        # )
         return type(self) == type(other) and self.args == other.args
 
-    def __getstate__(self) -> Dict[str, Any]:
-        return {key: getattr(self, key) for key in self.argnames}
+    def __hash__(self):
+        return self._hash
 
-    @property
-    def args(self):
-        if (result := self._args) is None:
-            result = self._args = tuple(
-                getattr(self, key) for key in self.argnames
-            )
-        return result
-
-    def __setstate__(self, state: Dict[str, Any]) -> None:
-        """Set state after unpickling.
-
-        Parameters
-        ----------
-        state
-            A dictionary storing the objects attributes.
-        """
-        self._args = None
-        for key, value in state.items():
-            setattr(self, key, value)
+    def __setattr__(self, name: str, _: Any) -> None:
+        raise TypeError(
+            f"{self.__class__.__name__} attribute {name!r} is read-only"
```

Comment:
Ideally the message indicates that the entire object is immutable, not just a specific property.